storage/replica.go

```diff
@@ -607,47 +740,62 @@ func (r *Replica) addAdminCmd(ctx context.Context, args proto.Request) (proto.Re
 // addReadOnlyCmd updates the read timestamp cache and waits for any
 // overlapping writes currently processing through Raft ahead of us to
 // clear via the read queue.
-func (r *Replica) addReadOnlyCmd(ctx context.Context, args proto.Request) (proto.Response, error) {
-	header := args.Header()
+func (r *Replica) addReadOnlyCmd(ctx context.Context, bArgs *proto.BatchRequest) (*proto.BatchResponse, error) {
+	header := bArgs.Header()
 
 	if err := r.checkCmdHeader(header); err != nil {
 		return nil, err
 	}
-
-	// If read-consistency is set to INCONSISTENT, run directly.
-	if header.ReadConsistency == proto.INCONSISTENT {
-		// But disallow any inconsistent reads within txns.
-		if header.Txn != nil {
-			return nil, util.Errorf("cannot allow inconsistent reads within a transaction")
-		}
-		if header.Timestamp.Equal(proto.ZeroTimestamp) {
-			header.Timestamp = r.rm.Clock().Now()
-		}
-		reply, intents, err := r.executeCmd(r.rm.Engine(), nil, args)
-		r.handleSkippedIntents(args, intents) // even on error
-		return reply, err
-	} else if header.ReadConsistency == proto.CONSENSUS {
-		return nil, util.Errorf("consensus reads not implemented")
+	if err := r.checkBatchRequest(bArgs); err != nil {
+		return nil, err
 	}
 
+	trace := tracer.FromCtx(ctx)
+
 	// Add the read to the command queue to gate subsequent
 	// overlapping commands until this command completes.
-	cmdKey := r.beginCmd(header, true)
-
-	// This replica must have leader lease to process a consistent read.
-	if err := r.redirectOnOrAcquireLeaderLease(tracer.FromCtx(ctx), header.Timestamp); err != nil {
-		r.endCmd(cmdKey, args, err, true /* readOnly */)
+	qDone := trace.Epoch("command queue")
+	cmdKeys, err := r.beginCmds(bArgs)
+	qDone()
+	if err != nil {
 		return nil, err
 	}
 
-	// Execute read-only command.
-	reply, intents, err := r.executeCmd(r.rm.Engine(), nil, args)
+	// If there are command keys (there might not be if reads are
+	// inconsistent), the read requires the leader lease.
+	if len(cmdKeys) > 0 {
+		if err := r.redirectOnOrAcquireLeaderLease(tracer.FromCtx(ctx), header.Timestamp); err != nil {
+			r.endCmds(cmdKeys, bArgs, err)
+			return nil, err
+		}
+	}
 
-	// Only update the timestamp cache if the command succeeded.
-	r.endCmd(cmdKey, args, err, true /* readOnly */)
+	// Execute read-only batch command.
+	bReply := &proto.BatchResponse{}
+	bReply.Timestamp = bArgs.Timestamp
+	var rErr error
```

Comment:
this is fragile, because there is already an `err` in scope. instead of doing this, can you `var` `reply` and `intents` inside the for-expression and assign to `err`?